internal/interceptor.go

```diff
@@ -0,0 +1,318 @@
+// The MIT License
+//
+// Copyright (c) 2021 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"time"
+
+	"github.com/uber-go/tally/v4"
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/sdk/converter"
+	"go.temporal.io/sdk/log"
+)
+
+// Interceptor is a common interface for all interceptors. See documentation in
+// the interceptor package for more details.
+type Interceptor interface {
+	ClientInterceptor
+	WorkerInterceptor
+}
+
+// WorkerInterceptor is a common interface for all interceptors. See
+// documentation in the interceptor package for more details.
+type WorkerInterceptor interface {
```

Comment:
Is this used as a factory/provider to create activity/workflow interceptor? Should the name be InboundInterceptorProvider?

Reply:
Why? It is an interceptor, conceptually, whether it provides other interceptors or directly intercepts. This matches other SDK naming _and_ existing Go SDK naming. This was intentionally changed from `WorkflowInterceptorFactory` to `WorkerInterceptor` over a year ago in #169 and this naming was approved in the proposal at https://github.com/temporalio/proposals/pull/45.